<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Fluseq by hkailee</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Fluseq</h1>
        <h2>A bioinformatics pipeline for influenza virus genome sequencing that includes contig assembly, blastn, filter for in-run, between-sample contamination reads, and consensus sequence calling.</h2>
        <a href="https://github.com/hkailee/FluSeq" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a id="fluseq-v10" class="anchor" href="#fluseq-v10" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>FluSeq-v1.0</h1>

<p>Version 1.0 for Illumina Miseq system.
For Research Use only. Not for use in diagnostic procedures.</p>

<p>FluSeq is a bioinformatics program developed and tested for an influenza A genome-wide amplicon-based high-throughput sequencing (HTS) method, using Illumina Miseq system [1]. It includes contigs assembly, blastn, filters for in-run contamination reads, and consensus sequence calling. Users are advised to refer the end-to-end laboratory protocol described by Lee, et al. 2016 [1]. Users are encouraged to modify the FluSeq program written in python scripts for other clinical virus amplicon-based HTS methods.</p>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>INSTALLATION</h2>

<p>Installation instructions are available for LINUX/UNIX or MacOSX only. The entire analytic workflow was implemented and tested by the author on CentOS-6.6/RedHat Linux and MacOSX, but not on other operating systems. The operating system should contain the latest version of java.</p>

<h3>
<a id="install-python-343-and-pip" class="anchor" href="#install-python-343-and-pip" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install Python 3.4.3 and pip</h3>

<p>(<a href="https://www.python.org/downloads/">https://www.python.org/downloads/</a>)</p>

<ul>
<li>Upon installation, log in as root from Terminal (Linux and MacOSX): </li>
</ul>

<pre><code># pip3.4 install pandas
# pip3.4 install numpy
# pip3.4 install ZODB
# pip3.4 install lxml 
# pip3.4 install xlrd
# pip3.4 install xlwt
# pip3.4 install beautifulsoup4
# pip3.4 install scipy (# yum install lapack-devel if lapack/blas resources not found in the system)
# pip3.4 install transaction
</code></pre>

<h3>
<a id="install-vicuna-v13" class="anchor" href="#install-vicuna-v13" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install VICUNA-v1.3</h3>

<p>(<a href="http://www.broadinstitute.org/scientific-community/science/projects/viral-genomics/vicuna">http://www.broadinstitute.org/scientific-community/science/projects/viral-genomics/vicuna</a>) </p>

<ul>
<li>The vicuna_config.txt used by this analytic workflow is stored in FluSeq Folder. Compilation of vicunAnalysis is not required during the installation. </li>
<li>Path to the compiled vicuna executive file needs to be changed accordingly in the FluSeq-v1.0.py, i.e. Line 100: First argument of the  subprocess.Popen, to “YourPathInstalled/VICUNA_v1.3/bin/vicuna-omp-v1.0”</li>
<li>Optional: Line 104 - logging.info: Change program path to “YourPathInstalled/VICUNA_v1.3/bin/vicuna-omp-v1.0”.</li>
</ul>

<h3>
<a id="install-blast-2231-software" class="anchor" href="#install-blast-2231-software" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install BLAST 2.2.31+ software</h3>

<p>(<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download</a>)</p>

<h3>
<a id="install-bwa-062-software" class="anchor" href="#install-bwa-062-software" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install BWA-0.6.2 software</h3>

<p>(<a href="http://sourceforge.net/projects/bio-bwa/files/">http://sourceforge.net/projects/bio-bwa/files/</a>)</p>

<ul>
<li>Users are advised to install BWA-0.6.2 software but not any version after this, as bwa-aln/sampe in this version was found to be more stable in reads alignment for this study.</li>
</ul>

<h3>
<a id="install-samtools-12" class="anchor" href="#install-samtools-12" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install samtools-1.2</h3>

<p>(<a href="http://sourceforge.net/projects/samtools/files/samtools/1.2/">http://sourceforge.net/projects/samtools/files/samtools/1.2/</a>)</p>

<h3>
<a id="install-gatk-34-46-software" class="anchor" href="#install-gatk-34-46-software" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install GATK-3.4-46 software</h3>

<p>(<a href="https://www.broadinstitute.org/gatk/download/">https://www.broadinstitute.org/gatk/download/</a>)</p>

<ul>
<li>Path to the GenomeAnalysisTK.jar executive file needs to be changed accordingly in the FluSeq-v1.0.py, i.e. Lines 233, 259,  and 269: Third or Fourth argument of the  subprocess.Popen, to “YourPathInstalled/GenomeAnalysisTK.jar”.</li>
<li>Optional: Lines 243, 265,  and 278: logging.info: Change program path to “YourPathInstalled/GenomeAnalysisTK.jar”.</li>
</ul>

<h3>
<a id="install-picard-tools-1138" class="anchor" href="#install-picard-tools-1138" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install picard-tools-1.138</h3>

<p>(<a href="https://github.com/broadinstitute/picard/releases/tag/1.138">https://github.com/broadinstitute/picard/releases/tag/1.138</a>)</p>

<ul>
<li>Path to the GenomeAnalysisTK.jar executive file needs to be changed accordingly in the FluSeq-v1.0.py, i.e. Line 146: Third argument of the  subprocess.Popen, to “YourPathInstalled/picard-tools-1.138/picard.jar”.</li>
<li>Optional: Line 152: logging.info: Change program path to “YourPathInstalled/picard-tools-1.138/picard.jar”.</li>
</ul>

<h3>
<a id="install-fluseq" class="anchor" href="#install-fluseq" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install FluSeq</h3>

<p>a.  Download the FluSeq.tar, decompress, and mv the FluSeq Folder to ~.</p>

<pre><code>$ tar –xvf FluSeq.tar
$ mv FluSeq ~ 
</code></pre>

<p>b.  Change working directory to FluSeq </p>

<pre><code>$ cd ~/FluSeq
</code></pre>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>USAGE</h2>

<h3>
<a id="influenza-sequence-database" class="anchor" href="#influenza-sequence-database" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Influenza Sequence Database</h3>

<p>This section describes the sequence database used for blastn. It contains all influenza sequences available in GenBank, updated to the date accessed by the user.</p>

<h4>
<a id="creatingupdating-sequence-database" class="anchor" href="#creatingupdating-sequence-database" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Creating/Updating Sequence Database</h4>

<ol>
<li> Download all influenza sequences available from NCBI Influenza Virus Resource with a customized FASTA defline as “&gt;{accession}|{strain}|{segment}|{serotype}” and save the fasta file as FASTA.fa in FASTADatabase directory.</li>
<li> Change working directory to FASTADatabase</li>
</ol>

<pre><code>$ cd ~/FluSeq/FASTADatabase
</code></pre>

<ol>
<li> Execute DBupdate</li>
</ol>

<pre><code>$ ./DBupdate.py FASTA.fa
or
$ python3 DBupdate.py FASTA.fa
or
$ python3.4 DBupdate.py FASTA.fa
</code></pre>

<h3>
<a id="fluseq-input-files-and-folder" class="anchor" href="#fluseq-input-files-and-folder" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>FluSeq INPUT FILES and FOLDER</h3>

<p>This section describes the input FluSeq-v1.0 requires.</p>

<h4>
<a id="folder-and-file-naming" class="anchor" href="#folder-and-file-naming" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Folder and File Naming</h4>

<p>The top-level input folder can be named in any combination of alphanumeric characters. It is advisable that the first letter of the name of a folder is in upper case for a better folder and file organization, e.g. Folder20151231 (<strong>Figure 1</strong> – Folder in red).</p>

<p>The input folder should contain paired-end reads in fastq.gz format generated by Miseq Reporter or fastq2bcl2, e.g. SampleID_L001_R1_001.fastq.gz and SampleID_L001_R2_001.fastq.gz (<strong>Figure 1</strong>). Also, it should contain the ResequencingRunStatistics.xml that can be copied from the top-level run folder named according to Miseq , e.g. YYMMDD_machinename_experimentnumber_flowcellnumber.</p>

<p><img src="https://cloud.githubusercontent.com/assets/13162113/13169042/5dc99350-d71d-11e5-9561-8402acce20e0.png" alt="Figure 1"></p>

<p><strong>Figure 1.</strong> Directory of FluSeq and input folder. Input folders for run analysis and contamination learning are colored in red and orange, respectively.</p>

<h3>
<a id="contamination-learning-optional" class="anchor" href="#contamination-learning-optional" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contamination Learning (Optional)</h3>

<p>The current FluSeq software provides the database file for MS2 contamination statistics (i.e. ms2SysCtrl.db). However, it is recommeded that individual laboratories generate a laboratory-specific database file to allow more accurate statistics. </p>

<p>The input folder for in-run contamination learning should contain paired-end reads for an in-run negative system control (NSC) in fastq.gz format generated by Miseq Reporter or fastq2bcl2, e.g. SampleID_L001_R1_001.fastq.gz and SampleID_L001_R2_001.fastq.gz (<strong>Figure 1</strong> – Folder in orange). Also, it should contain the ResequencingRunStatistics.xml that can be copied from the top-level run folder named according to Miseq , e.g. YYMMDD_machinename_experimentnumber_flowcellnumber. </p>

<p>The database will be updated according for each execution. A new database file will be generated if the ms2SysCtrl.db file is not found in the FluSeq folder.</p>

<h4>
<a id="procedure" class="anchor" href="#procedure" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Procedure</h4>

<ol>
<li> Change working directory to FluSeq </li>
</ol>

<pre><code>$ cd ~/FluSeq
</code></pre>

<ol>
<li> Execute FluSeq</li>
</ol>

<pre><code>$ ./ms2ContLearning.py InputFolder 151206
or
$ python3 ms2ContLearning.py InputFolder 151206
or
$ python3.4 ms2ContLearning.py InputFolder 151206
</code></pre>

<h3>
<a id="execution-of-fluseq-analysis" class="anchor" href="#execution-of-fluseq-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Execution of FluSeq analysis</h3>

<p>a.  Change working directory to FluSeq </p>

<pre><code>$ cd ~/FluSeq
</code></pre>

<p>b.  Execute FluSeq</p>

<pre><code>$ ./FluSeq-v1.0.py InputFolder
or
$ python3 FluSeq-v1.0.py InputFolder
or
$ python3.4 FluSeq-v1.0.py InputFolder
</code></pre>

<h2>
<a id="reference" class="anchor" href="#reference" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>REFERENCE:</h2>

<p>[1] Hong Kai Lee, Chun Kiat Lee, Julian Wei-Tze Tang, Tze Ping Loh, and Evelyn Siew-Chuan Koay. Contamination-controlled high-throughput whole genome sequencing for influenza A viruses using the MiSeq sequencer. <em>Sci Rep.</em> 2016, 6:33318. doi: 10.1038/srep33318.</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/hkailee/FluSeq/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/hkailee/FluSeq/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/hkailee/FluSeq"></a> is maintained by <a href="https://github.com/hkailee">hkailee</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-71299750-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
