<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Fluseq by hkailee</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Fluseq</h1>
      <h2 class="project-tagline">A bioinformatics program for influenza sequencing that include contig assembly, blastn, filter for contamination reads, and consensus sequence calling.</h2>
      <a href="https://github.com/hkailee/FluSeq" class="btn">View on GitHub</a>
      <a href="https://github.com/hkailee/FluSeq/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/hkailee/FluSeq/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h3>
<a id="welcome-to-fluseq-page" class="anchor" href="#welcome-to-fluseq-page" aria-hidden="true"><span class="octicon octicon-link"></span></a>Welcome to FluSeq Page.</h3>

<p>FluSeq is a bioinformatics program developed and tested for an influenza A genome-wide amplicon-based HTS method, using Illumina Miseq system [1]. It includes contig assembly, blastn, filters for contamination reads, and consensus sequence calling. Users are advised to refer the end-to-end laboratory protocol described by Lee, et al. 2016 [1]. The users are encouraged to modified the FluSeq program written in simple python scripts for other clinical virus amplicon-based HTS method.</p>

<h3>
<a id="reference" class="anchor" href="#reference" aria-hidden="true"><span class="octicon octicon-link"></span></a>Reference:</h3>

<p>[1] Hong Kai Lee, Chun Kiat Lee, Julian Wei-Tze Tang, Tze Ping Loh, and Evelyn Siew-Chuan Koay. Pairwise comparison of contamination-controlled high-throughput and Sanger sequencing for influenza A genome sequencing. [Article in preparation]</p>

<h3>
<a id="authors-and-contributors" class="anchor" href="#authors-and-contributors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>In 2016, HK Lee (<a href="https://github.com/hkailee" class="user-mention">@hkailee</a>) developed FluSeq.</p>

<h3>
<a id="support-or-contact" class="anchor" href="#support-or-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support or Contact</h3>

<p>The users are encouraged to contact author regarding FluSeq-related technical issues. <a href="https://github.com/contact">contact support</a>.</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/hkailee/FluSeq">Fluseq</a> is maintained by <a href="https://github.com/hkailee">hkailee</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-71299750-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
